src/analysis/calculus/mean_value.lean

```diff
@@ -42,84 +42,177 @@ set_option class.instance_max_depth 120
-    show bdd_above K, from ⟨1, λy hy, hy.1.2⟩,
-    have A : continuous_on (λt:ℝ, (∥f t - f 0∥, D * t)) (Icc 0 1),
-    { apply continuous_on.prod,
-      { refine continuous_norm.comp_continuous_on _,
-        apply continuous_on.sub hf.continuous_on continuous_on_const },
-      { exact (continuous_const.mul continuous_id).continuous_on } },
-    show is_closed K, from
-      A.preimage_closed_of_closed is_closed_Icc (ordered_topology.is_closed_le' _) },
-  have : k = 1,
-  { by_contradiction k_ne_1,
-    have k_lt_1 : k < 1 := lt_of_le_of_ne k_mem_K.1.2 k_ne_1,
-    have : 0 ≤ k := k_mem_K.1.1,
-    let g := deriv_within f (Icc 0 1) k,
-    let h := λx, f x - f k - (x-k) • g,
-    have : is_o (λ x, h x) (λ x, x - k) (nhds_within k (Icc 0 1)) :=
-      (hf k k_mem_K.1).has_deriv_within_at,
-    have : {x | ∥h x∥ ≤ (D-C) * ∥x-k∥} ∈ nhds_within k (Icc 0 1) :=
-      this (D-C) (sub_pos_of_lt hD),
-    rcases mem_nhds_within.1 this with ⟨s, s_open, ks, hs⟩,
-    rcases is_open_iff.1 s_open k ks with ⟨ε, εpos, hε⟩,
-    change 0 < ε at εpos,
-    let δ := min ε (1-k),
-    have δpos : 0 < δ, by simp [δ, εpos, k_lt_1],
-    let x := k + δ/2,
-    have k_lt_x : k < x, by { simp only [x], linarith },
-    have x_lt_1 : x < 1 := calc
-      x < k + δ : add_lt_add_left (half_lt_self δpos) _
-      ... ≤ k + (1-k) : add_le_add_left (min_le_right _ _) _
-      ... = 1 : by ring,
-    have xε : x ∈ ball k ε,
-    { simp [dist, x, abs_of_nonneg (le_of_lt (half_pos δpos))],
-      exact lt_of_lt_of_le (half_lt_self δpos) (min_le_left _ _) },
-    have xI : x ∈ Icc (0:ℝ) 1 :=
-      ⟨le_of_lt (lt_of_le_of_lt (k_mem_K.1.1) k_lt_x), le_of_lt x_lt_1⟩,
-    have Ih : ∥h x∥ ≤ (D - C) * ∥x - k∥ :=
-      hs ⟨hε xε, xI⟩,
-    have I : ∥f x - f k∥ ≤ D * (x-k) := calc
-      ∥f x - f k∥ = ∥(x-k) • g + h x∥ : by { congr' 1, simp only [h], abel }
-      ... ≤ ∥g∥ * ∥x-k∥ + (D-C) * ∥x-k∥ : norm_add_le_of_le (by rw [norm_smul, mul_comm]) Ih
-      ... ≤ C * ∥x-k∥ + (D-C) * ∥x-k∥ :
-        add_le_add_right (mul_le_mul_of_nonneg_right (bound k k_mem_K.1) (norm_nonneg _)) _
-      ... = D * ∥x-k∥ : by ring
-      ... = D * (x-k) : by simp [norm, abs_of_nonneg (le_of_lt (half_pos δpos))],
-    have : ∥f x - f 0∥ ≤ D * x := calc
-      ∥f x - f 0∥ = ∥(f x - f k) + (f k - f 0)∥ : by { congr' 1, abel }
-      ... ≤ D * (x - k) + D * k : norm_add_le_of_le I (k_mem_K.2)
-      ... = D * x : by ring,
-    have xK : x ∈ K := ⟨xI, this⟩,
-    have : x ≤ k := le_cSup ⟨1, λy hy, hy.1.2⟩ xK,
-    exact (not_le_of_lt k_lt_x) this },
-  rw this at k_mem_K,
-  simpa [this] using k_mem_K.2
+  refine image_norm_le_of_norm_limsup_slope_right_le_deriv_boundary hf _ ha hB hB' bound,
+  assume x hx,
+  convert tendsto_le_right _
+    ((has_deriv_within_at_iff_tendsto_slope' $ lt_irrefl x).1 (hf' x hx)).norm,
+  { ext z, rw [norm_smul, normed_field.norm_inv] },
+  rw (nhds_eq_orderable ∥f' x∥),
+  exact inf_le_right
 end
 
+lemma image_norm_le_of_norm_deriv_right_le_deriv_boundary {f' : ℝ → E}
```

Comment:
You could also give a lemma with the large inequality in `bound`, i.e., replacing it with `(bound : ∀ x ∈ Ico a b, ∥f' x∥ ≤ deriv B x)`. It follows by applying your version to `B x + epsilon (x-a)` and letting `epsilon` tend to `0`.

Reply:
Done.